features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/metric/SessionEndedMetric.kt

```diff
@@ -0,0 +1,222 @@
+) {
+
+    private val trackedViewMap by lazy { mutableMapOf<String, TrackedView>() }
+
+    private val trackedErrorMap by lazy { mutableMapOf<String, Int>() }
+
+    private var firstTrackedView: TrackedView? = null
+    private var lastTrackedView: TrackedView? = null
+    private var wasStopped: Boolean = false
+
+    /**
+     * Called on view tracked event, return true if the view is recorded by the metric, false otherwise.
+     */
+    fun onViewTracked(rumViewEvent: ViewEvent): Boolean {
+        if (rumViewEvent.session.id != sessionId) {
+            return false
+        }
+        val trackedView = trackedViewMap[rumViewEvent.view.id] ?: TrackedView(
+            viewUrl = rumViewEvent.view.url,
+            startMs = rumViewEvent.date,
+            durationNs = rumViewEvent.view.timeSpent
+        )
+
+        trackedViewMap[rumViewEvent.view.id] = trackedView.copy(durationNs = rumViewEvent.view.timeSpent)
+        if (firstTrackedView == null) {
+            firstTrackedView = trackedView
+        }
+        lastTrackedView = trackedView
+        return true
+    }
+
+    fun onErrorTracked(sdkErrorKind: String) {
+        trackedErrorMap[sdkErrorKind] = (trackedErrorMap[sdkErrorKind] ?: 0) + 1
+    }
+
+    fun onSessionStopped() {
+        wasStopped = true
+    }
+
+    fun toMetricAttributes(): Map<String, Any?> {
+        return resolveAttributes()
+    }
+
+    private fun resolveAttributes(): Map<String, Any?> {
+        return mapOf(
+            METRIC_TYPE_KEY to METRIC_TYPE_VALUE,
+            RSE_KEY to resolveRseAttributes()
+        )
+    }
+
+    private fun calculateDuration(): Long {
+        return lastTrackedView?.let { last ->
+            firstTrackedView?.let { first ->
+                TimeUnit.MILLISECONDS.toNanos(last.startMs - first.startMs) + last.durationNs
+            }
+        } ?: 0L
+    }
+
+    private fun resolveRseAttributes(): Map<String, Any?> {
+        return mapOf(
+            PROCESS_TYPE_KEY to PROCESS_TYPE_VALUE,
+            PRECONDITION_KEY to precondition,
+            DURATION_KEY to calculateDuration(),
+            WAS_STOPPED_KEY to wasStopped,
+            VIEW_COUNTS_KEY to resolveViewCountsAttributes(),
+            SDK_ERRORS_COUNT_KEY to resolveSDKErrorsCountAttributes()
+        )
+    }
+
+    private fun resolveViewCountsAttributes(): Map<String, Any?> {
+        return mapOf(
+            VIEW_COUNTS_TOTAL_KEY
+                to trackedViewMap.size,
+            VIEW_COUNTS_BG_KEY
+                to trackedViewMap.values.count { it.viewUrl == RumViewManagerScope.RUM_BACKGROUND_VIEW_URL },
+            VIEW_COUNTS_APP_LAUNCH_KEY
+                to trackedViewMap.values.count { it.viewUrl == RumViewManagerScope.RUM_APP_LAUNCH_VIEW_URL }
+        )
+    }
+
+    private fun resolveSDKErrorsCountAttributes(): Map<String, Any?> {
+        return mapOf(
+            SDK_ERRORS_COUNT_TOTAL_KEY to trackedErrorMap.values.sumOf {
+                it
+            },
```

Comment:
nit: can be also like that, if I'm not wrong

```suggestion
            SDK_ERRORS_COUNT_TOTAL_KEY to trackedErrorMap.values.sum(),
```

Reply:
updated